### ISSUE TITLE
Huge Page with Fork

### ISSUE THREAD

**Issue by alphanso** [gperftools/gperftools#1139]
Hi,

I am dynamically linking tcmalloc by linking tcmalloc.so at compile time. I am trying to use 1GB huge pages with tcmalloc. We have one fork at start of the process. It causes issue in our program and memory gets corrupted. 

Commands to reproduce bugs: 
$ g++ -std=c++14 sample_program.cpp -o sample.o -lpthread -ltcmalloc_debug
$ TCMALLOC_MEMFS_MALLOC_PATH=/dev/hugepages/ ./sample.o

[sample_program.cpp.txt'](https://github.com/gperftools/gperftools/files/3550106/sample_program.cpp.txt)

**Comment by alokkataria** [comment]
I think the issue is that when tcmalloc uses hugepages, after the allocation the child process still maps to the same physical addresses. So any writes from the child/parent end up overwriting the same location. Without hugepages though a write from child/parent ends up doing a CopyOnWrite and allocating new memory so things work fine there.
I don't know if there's a way to change that behaviour with hugepages though.  Can someone familiar with the tcmalloc hugepage allocator confirm this behaviour ?

**Comment by alokkataria** [comment]
BTW, here is a simple program that depicts the behaviour I mentioned in the previous comment. This program does following things 
1. Allocates 10 objects of 2MB each
2. The does a fork, from the parent overwrites the values at the allocated locations. 
3. From the child verifies that the child still sees the old values. 

Note that without hugepage allocator it shows the expected behaviour but the check fails when Hugepage are being used.


[tcmalloc-test.cc.txt](https://github.com/gperftools/gperftools/files/3572937/tcmalloc-test.cc.txt)
[tcmalloc-test.cc.txt](https://github.com/gperftools/gperftools/files/3572938/tcmalloc-test.cc.txt)

**Comment by alokkataria** [comment]
@alphanso you might want to set TCMALLOC_MEMFS_MAP_PRIVATE=true, this ensures mapping are not shared across parent and child.

**Comment by alk** [comment]
Indeed. Thanks. Also private should be default. Would be happy to apply patch that fixes that.
